Sources/NIOSOCKS/State/ClientStateMachine.swift

```diff
@@ -0,0 +1,151 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+enum ClientState: Hashable {
+    case inactive
+    case waitingForClientGreeting
+    case waitingForAuthenticationMethod(ClientGreeting)
+    case waitingForClientRequest
+    case waitingForServerResponse(ClientRequest)
+    case active
+    case error
+}
+
+enum ClientAction: Hashable {
+    case waitForMoreData
+    case sendGreeting
+    case sendRequest
+    case proxyEstablished
+    case sendData(ByteBuffer)
+}
+
+struct ClientStateMachine {
+
+    private var state: ClientState
+    
+    var proxyEstablished: Bool {
+        switch self.state {
+        case .active:
+            return true
+        default:
```

Comment:
As a general stylistic note I tend to prefer to list all the states explicitly in these `switch`es. This forces me to re-evaluate these each time I add a new state. This is more of a suggestion than a strong piece of feedback though.